### PR TITLE
DKG Finalization Optimizations

### LINKED ISSUE #2536: [cryptography/dkg] Add `finalize_trusted`
Some applications may wish to verify dealings included in blocks to minimize delay at the end of an application epoch.

We should explore a way to provide pre-verified dealer logs to finalize to avoid duplicate verification work.

### LINKED ISSUE #3212: [cryptography] DKG: Low hanging optimization fruit
- We can use batch ed25519 verification in select,
- We can batch other share checks.

### PR DESCRIPTION
**Pull request by cronokirby**
Stacked on top of #3312.

Closes #2536.
Closes #3212.
Related https://github.com/commonwarexyz/monorepo/issues/1811.

This adds a new structure for containing dealer logs, allowing them to be verified in advance.

We also implement optimizations for a given dealer, batching signature checks, and reveal checks.

One big win is implementing parallelism to check multiple dealers at once (while preserving the order we select dealers in), which allows the method to actually scale, whereas it was bottlenecked previously.

### CODE REVIEW

File: cryptography/src/bls12381/dkg.rs [hunk 1]
```diff
@@ -2429,7 +2586,13 @@ mod test_plan {
...
 
                 // Make sure that bad dealers are not selected.
-                let selection = select::<_, _, N3f1>(&info, dealer_logs.clone());
+                let mut logs = Logs::<_, _, N3f1>::default();
+                for (dealer, log) in &dealer_logs {
+                    logs.record(dealer.clone(), log.clone());
+                }
+                let selection = logs
+                    .clone()
+                    .select::<ed25519::Batch>(&info, &mut rng, &Sequential);
```
[the remark this answers is not in the document]
Comment (cronokirby) [reply cryptography/src/bls12381/dkg.rs]: It's still deterministic though, even if "fragile" which is fine, it's random.

**Comment by patrick-ogrady** [comment]
Not sure if this is an issue yet (but seems like we should have more testing around this edge case):

```
  - [P1] Preserve dealer-order selection after incremental pre_verify — /Users/patrickogrady/code/monorepo-pr-3328/cryptography/src/bls12381/
    dkg.rs:1383-1385
    pre_verify() only checks max(needed, unknown_in_required_prefix) unknown dealers, so if some low-sorted dealers were already cached as
    invalid and enough later dealers are cached as valid, a newly recorded valid dealer can remain None in known even though its key sorts before
    one of those later dealers. select() then skips that dealer and takes the later one, which makes the chosen subset depend on arrival/pre-
    verification history instead of dealer order. In the initial DKG this changes the summed public polynomial, and in resharing it still changes
    Output.dealers, so honest participants can derive different outputs from the same final log set.
```

**Comment by cronokirby** [comment]
> Not sure if this is an issue yet (but seems like we should have more testing around this edge case):
> 
> ```
>   - [P1] Preserve dealer-order selection after incremental pre_verify — /Users/patrickogrady/code/monorepo-pr-3328/cryptography/src/bls12381/
>     dkg.rs:1383-1385
>     pre_verify() only checks max(needed, unknown_in_required_prefix) unknown dealers, so if some low-sorted dealers were already cached as
>     invalid and enough later dealers are cached as valid, a newly recorded valid dealer can remain None in known even though its key sorts before
>     one of those later dealers. select() then skips that dealer and takes the later one, which makes the chosen subset depend on arrival/pre-
>     verification history instead of dealer order. In the initial DKG this changes the summed public polynomial, and in resharing it still changes
>     Output.dealers, so honest participants can derive different outputs from the same final log set.
> ```

Yeah, this was an issue, pushed a fix and a regression test.

File: examples/reshare/src/dkg/actor.rs
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs:476]: Now that `summary` is in `Logs`, wondering if we should also keep it inside this data structure as well (so it just becomes `log.check()`)?

Or, maybe that move into `Logs` was a bit overzealous (just got a little annoyed by how many places we passed around summary).

Comment (cronokirby) [reply examples/reshare/src/dkg/actor.rs:476]: Moving Info into `Logs` seems like a fine move to me. I'm not sure I understand what data structure this comment refers to though.

Comment (patrick-ogrady) [reply examples/reshare/src/dkg/actor.rs:476]: This comment doesn't make sense. Disregard.

File: examples/reshare/src/dkg/actor.rs [hunk 2]
```diff
@@ -514,10 +516,17 @@ where
...
+                        let mut logs = Logs::<_, _, N3f1>::new(round.clone());
+                        for (dealer, log) in storage.logs(epoch) {
+                            logs.record(dealer, log);
+                        }
                         let (success, next_round, next_output, next_share) =
                             if let Some(ps) = player_state.take() {
-                                match ps.finalize::<N3f1>(logs, &Sequential) {
+                                match ps.finalize::<N3f1, Batch>(
+                                    logs,
+                                    &mut self.context,
```
Comment (patrick-ogrady) [inline examples/reshare/src/dkg/actor.rs:528]: nit: move context to first arg

File: cryptography/src/bls12381/dkg.rs [hunk 3]
```diff
@@ -1226,6 +1293,155 @@ fn transcript_for_log<V: Variant, P: PublicKey>(
...
+    pub fn pre_verify<B: BatchVerifier<PublicKey = P>>(
+        &mut self,
+        rng: &mut impl CryptoRngCore,
+        strategy: &impl Strategy,
+    ) {
+        let required_commitments = self.info.required_commitments::<M>() as usize;
+        let transcript = transcript_for_round(&self.info);
+
+        // Create a pending batch, which we try and optimistically size as small as
+        // possible, to avoid verifying more dealers than we need, if they're all
```
Comment (patrick-ogrady) [inline cryptography/src/bls12381/dkg.rs:1375]: Someday, we'll switch to the "first 2f+1" 🙃

File: cryptography/src/bls12381/dkg.rs [hunk 4]
```diff
@@ -1226,6 +1293,155 @@ fn transcript_for_log<V: Variant, P: PublicKey>(
...
+
+    /// Given the logs we've received, determine which dealer logs to use, if any.
+    ///
+    /// This might return an error if there are not enough good logs that we can use.
+    fn select<B: BatchVerifier<PublicKey = P>>(
+        mut self,
+        rng: &mut impl CryptoRngCore,
+        strategy: &impl Strategy,
+    ) -> Result<SelectedLogs<V, P>, Error> {
+        self.pre_verify::<B>(rng, strategy);
```
Comment (patrick-ogrady) [inline cryptography/src/bls12381/dkg.rs:1430]: ❤️

**Review by patrick-ogrady — APPROVED** [review]
